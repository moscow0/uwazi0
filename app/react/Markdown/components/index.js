import loadable from '@loadable/component';

import { Link } from 'react-router';
import { Icon } from 'UI';
import Counter from './Counter';
import ContactForm from './ContactForm';
import { EntityData } from './EntityData';
import { EntitySection } from './EntitySection';
import EntityLink from './EntityLink';
import ItemList from './ItemList';
import Repeat from './Repeat';
import Context from './Context';
import Connect from './Connect';
import Map from './Map';
import MarkdownLink from './MarkdownLink';
import MarkdownMedia from './MarkdownMedia';
import PayPalDonateLink from './PayPalDonateLink';
import PublicForm from './PublicForm';
import Value from './Value';
import SearchBox from './SearchBox';
import EntityInfo from './EntityInfo';

const BarChart = loadable(async () =>
  import(/* webpackChunkName: "LazyLoadBarChart" */ './BarChart.js')
);
const FreeBarChart = loadable(async () =>
  import(/* webpackChunkName: "LazyLoadFreeBarChart" */ './FreeBarChart.tsx')
);

const Slideshow = loadable(async () =>
  import(/* webpackChunkName: "LazyLoadSlideshow" */ './Slideshow.js')
);

const PieChart = loadable(async () =>
  import(/* webpackChunkName: "LazyLoadPieChart" */ './PieChart.js')
);

const ListChart = loadable(async () =>
  import(/* webpackChunkName: "LazyLoadListChart" */ './ListChart.js')
);
const GaugeChart = loadable(async () =>
  import(/* webpackChunkName: "LazyLoadGaugeChart" */ './GaugeChart.js')
);

export {
  MarkdownMedia,
  ContactForm,
  Context,
  Connect,
  EntityData,
  EntityLink,
  ItemList,
  Slideshow,
  MarkdownLink,
  PayPalDonateLink,
  PublicForm,
  SearchBox,
  Counter,
  BarChart,
  FreeBarChart,
  PieChart,
  ListChart,
  Repeat,
  GaugeChart,
  Value,
  Icon,
  Map,
  Link,
  EntityInfo,
  EntitySection,
};
