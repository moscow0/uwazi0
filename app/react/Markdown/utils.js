export const objectPath = (path, object) =>
  path.split('.').reduce((o, key) => {
    if (!o || !key) {
      return o;
    }
    return o.toJS ? o.get(key) : o[key];
  }, object);

export const logError = (err, propValueOf, propLabelOf) => {
  /* eslint-disable no-console */
  console.error('Error on EntityData: ');
  console.error('value-of: ', propValueOf, '; label-of: ', propLabelOf);
  console.error(err);
  /* eslint-enable no-console */
};
