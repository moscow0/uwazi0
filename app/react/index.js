import React from 'react';
import ReactDOM from 'react-dom';

import * as Sentry from '@sentry/react';

import { BrowserTracing } from '@sentry/tracing';

import { AppContainer } from 'react-hot-loader';

import App from './App';

import './App/sockets';

if (window.SENTRY_APP_DSN) {
  Sentry.init({
    release: window.UWAZI_VERSION,
    environment: window.UWAZI_ENVIRONMENT,
    dsn: window.SENTRY_APP_DSN,
    integrations: [new BrowserTracing()],
    tracesSampleRate: 0.1,
  });
}

const render = Component => {
  ReactDOM.hydrate(
    <AppContainer>
      <Component />
    </AppContainer>,
    document.getElementById('root')
  );
};

render(App);

if (module.hot) {
  module.hot.accept('./App', () => {
    const nextApp = require('./App');
    render(nextApp);
  });
}
