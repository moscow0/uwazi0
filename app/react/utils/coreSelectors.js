import { createSelector } from 'reselect';

export const selectTemplates = createSelector(
  s => s.templates,
  t => t.toJS()
);
