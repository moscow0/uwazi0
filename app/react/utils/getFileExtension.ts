export const getFileExtension = (filename: string) =>
  filename ? filename.substr(filename.lastIndexOf('.') + 1) : '';
