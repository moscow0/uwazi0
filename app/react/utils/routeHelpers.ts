export function validateHomePageRoute(route: string) {
  const queryParams = '(\\?.*)?)';
  const languageMatch = '(/[a-zA-Z]{2,3})?';
  const library = `(library(/map)?(/table)?/?${queryParams}`;
  const entity = '(entity/.+)';
  const page = '(page/.+)';
  const matcher = new RegExp(`${languageMatch}/(${library}|${page}|${entity})$`);
  return route.match(matcher) !== null;
}
