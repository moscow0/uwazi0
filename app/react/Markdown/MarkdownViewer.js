import PropTypes from 'prop-types';
import React, { Component } from 'react';
import rison from 'rison-node';
import { Translate } from 'app/I18N';
import { MarkdownLink, SearchBox, MarkdownMedia, ItemList } from './components';
import CustomHookComponents from './CustomHooks';

import markdownToReact from './markdownToReact';

class MarkdownViewer extends Component {
  static errorHtml(index, message) {
    return (
      <p key={index} className="error">
        <br />
        <strong>
          <i>
            <Translate translationKey="Custom component error">
              Custom component markup error: unsuported values! Please check your configuration
            </Translate>
          </i>
        </strong>
        <br />
        {message}
        <br />
      </p>
    );
  }

  static customHook(config, index) {
    const props = rison.decode(config);
    if (!CustomHookComponents[props.component]) {
      throw new Error('Invalid  component');
    }
    const Element = CustomHookComponents[props.component];
    return <Element {...props} key={index} />;
  }

  inlineComponent(type, config, index) {
    const { compact } = this.props;
    let result;
    if (type === 'list') {
      result = this.list(config, index);
    }

    if (type === 'link') {
      result = <MarkdownLink {...rison.decode(config)} key={index} />;
    }

    if (type === 'searchbox') {
      result = <SearchBox {...rison.decode(config)} key={index} />;
    }

    if (['vimeo', 'youtube', 'media'].includes(type)) {
      result = (
        <div key={index}>
          <MarkdownMedia key={index} config={config} compact={compact} />
        </div>
      );
    }

    if (type === 'customhook') {
      result = MarkdownViewer.customHook(config, index);
    }
    return result;
  }

  customComponent(type, config, index, children) {
    try {
      if (typeof type === 'function' || typeof type === 'object') {
        const Element = type;
        return (
          <Element {...config} key={index}>
            {children}
          </Element>
        );
      }

      if (type) {
        return this.inlineComponent(type, config, index);
      }
    } catch (error) {
      return MarkdownViewer.errorHtml(index, error.message);
    }

    return false;
  }

  list(_config, index) {
    const listData = this.props.lists[this.renderedLists] || {};
    const output = (
      <ItemList
        key={index}
        link={`/library/${listData.params}`}
        items={listData.items}
        options={listData.options}
      />
    );
    this.renderedLists += 1;
    return output;
  }

  render() {
    this.renderedLists = 0;

    const ReactFromMarkdown = markdownToReact(
      this.props.markdown,
      this.customComponent.bind(this),
      this.props.html
    );

    if (!ReactFromMarkdown) {
      return false;
    }

    return <div className="markdown-viewer">{ReactFromMarkdown}</div>;
  }
}

MarkdownViewer.defaultProps = {
  lists: [],
  markdown: '',
  html: false,
  compact: false,
};

MarkdownViewer.propTypes = {
  markdown: PropTypes.string,
  lists: PropTypes.arrayOf(PropTypes.object),
  html: PropTypes.bool,
  compact: PropTypes.bool,
};

export default MarkdownViewer;
