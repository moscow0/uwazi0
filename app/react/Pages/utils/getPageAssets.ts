import rison from 'rison-node';
import { get, has, uniq } from 'lodash';
import api from 'app/Search/SearchAPI';
import { markdownDatasets } from 'app/Markdown';
import { RequestParams } from 'app/utils/RequestParams';
import PagesAPI from '../PagesAPI';
import pageItemLists from './pageItemLists';

type Query = { filters: {}; types: string[]; limit?: string };

interface ListsData {
  params: string[];
  content: string;
  options: { limit?: number }[];
  searchs?: any;
}

const buildQuery = (sanitizedParams: string) => {
  const queryDefault = { filters: {}, types: [] };
  if (sanitizedParams) {
    const query = rison.decode(sanitizedParams.replace('?q=', '') || '()');
    if (typeof query !== 'object') {
      return queryDefault;
    }
    return query;
  }
  return queryDefault;
};

const prepareLists = (content: string, requestParams: RequestParams) => {
  const listsData: ListsData = pageItemLists.generate(content);

  listsData.searchs = Promise.all(
    listsData.params.map((params, index) => {
      const sanitizedParams = params ? decodeURI(params) : '';

      const query: Query = buildQuery(sanitizedParams);

      query.limit = listsData.options[index].limit ? String(listsData.options[index].limit) : '6';
      return api.search(requestParams.set(query));
    })
  );

  return listsData;
};

const replaceDynamicProperties = (pageContent?: string, datasets?: any) => {
  if (!pageContent || !datasets || (!datasets.entityData && !datasets.template)) {
    return { content: pageContent, errors: [] };
  }

  const parsableDatasets = {
    entity: datasets.entityData,
    template: datasets.template,
  };

  const errors: string[] = [];
  const content = pageContent.replace(/\$\{((entity.|template.)[^}^\s]*)\}/g, (match, p) => {
    switch (true) {
      case /entity.metadata.\w*$/.test(p):
        return get(parsableDatasets, `${p}[0].value`);

      case /entity.metadata.\w*.(value|displayValue)$/.test(p): {
        const path = p.split('.');
        const pathEnd = path.pop();
        return get(parsableDatasets, `${path.join('.')}[0].${pathEnd}`);
      }

      case /entity.metadata.\w*\[\d+]$/.test(p):
        return get(parsableDatasets, `${p}.value`);

      case has(parsableDatasets, p):
        return get(parsableDatasets, p);

      default:
        errors.push(match);
        return match;
    }
  });
  return { content, errors };
};

const getPageAssets = async (
  requestParams: RequestParams,
  additionalDatasets?: {},
  localDatasets?: {}
) => {
  const page = await PagesAPI.getById(requestParams);

  const { content, errors } = replaceDynamicProperties(page.metadata?.content, localDatasets);
  page.metadata.content = content;

  const listsData = prepareLists(page.metadata.content, requestParams);

  const dataSets = markdownDatasets.fetch(page.metadata.content, requestParams.onlyHeaders(), {
    additionalDatasets,
  });

  const [pageView, searchParams, searchOptions, datasets, listSearchs] = await Promise.all([
    page,
    listsData.params,
    listsData.options,
    dataSets,
    listsData.searchs,
  ]);
  pageView.scriptRendered = false;

  const itemLists = searchParams.map((p, index) => ({
    params: p,
    items: listSearchs[index].rows,
    options: searchOptions[index],
  }));

  const failedExpressions = uniq(errors).join('\n');
  return {
    pageView,
    itemLists,
    datasets: { ...datasets, ...localDatasets },
    ...(failedExpressions.length && {
      errors: `The following expressions are not valid properties:\n ${failedExpressions}`,
    }),
  };
};

export { getPageAssets };
