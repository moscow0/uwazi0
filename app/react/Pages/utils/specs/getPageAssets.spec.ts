/* eslint-disable no-template-curly-in-string */
import api from 'app/Search/SearchAPI';
import { RequestParams } from 'app/utils/RequestParams';
import { markdownDatasets } from 'app/Markdown';

import PagesAPI from '../../PagesAPI';
import pageItemLists from '../pageItemLists';

import { getPageAssets } from '../getPageAssets';

describe('getPageAssets', () => {
  let page: { _id: string; title: string; metadata: { content: string } };
  let data;
  let request: RequestParams;
  let apiSearch: jasmine.Spy;

  beforeEach(() => {
    page = {
      _id: 'abc2',
      title: 'Page 1',
      metadata: /*non-metadata-object*/ { content: 'originalContent' },
    };

    spyOn(PagesAPI, 'getById').and.callFake(async () => Promise.resolve(page));

    spyOn(pageItemLists, 'generate').and.returnValue({
      content: 'parsedContent',
      params: [
        '?q=(a:1,b:2)',
        '',
        '?q=(x:1,y:!(%27array%27),limit:24)',
        '?order=metadata.form&treatAs=number',
      ],
      options: [{}, { limit: 9 }, { limit: 0 }, { limit: 12 }],
    });

    let searchCalls = -1;

    apiSearch = jasmine.createSpy('apiSearch').and.callFake(async () => {
      searchCalls += 1;
      return Promise.resolve({ rows: [`resultsFor:${searchCalls}`] });
    });

    spyOn(api, 'search').and.callFake(apiSearch);

    data = { sharedId: 'abc2' };
    request = new RequestParams(data, 'headers');
  });

  it('should request page for view', async () => {
    const stateActions = await getPageAssets(request);

    expect(PagesAPI.getById).toHaveBeenCalledWith(request);
    expect(stateActions).toMatchSnapshot();
  });

  const assertItemLists = (itemLists: any) => {
    expect(itemLists.length).toBe(4);
    expect(itemLists[0].params).toBe('?q=(a:1,b:2)');
    expect(itemLists[0].items).toEqual(['resultsFor:0']);
    expect(itemLists[1].params).toBe('');
    expect(itemLists[1].items).toEqual(['resultsFor:1']);
    expect(itemLists[2].params).toBe('?q=(x:1,y:!(%27array%27),limit:24)');
    expect(itemLists[2].items).toEqual(['resultsFor:2']);
    expect(itemLists[3].params).toBe('?order=metadata.form&treatAs=number');
    expect(itemLists[3].items).toEqual(['resultsFor:3']);
    expect(itemLists[3].options).toEqual({ limit: 12 });
  };

  it('should request each list inside the content limited to 6 items (default) or the passed value and set the state', async () => {
    const stateActions = await getPageAssets(request);

    expect(apiSearch.calls.count()).toBe(4);
    expect(JSON.parse(JSON.stringify(apiSearch.calls.argsFor(0)[0]))).toEqual({
      data: { a: 1, b: 2, limit: '6' },
      headers: 'headers',
    });
    expect(apiSearch.calls.argsFor(1)[0]).toEqual({
      data: { filters: {}, types: [], limit: '9' },
      headers: 'headers',
    });

    expect(JSON.parse(JSON.stringify(apiSearch.calls.argsFor(2)[0]))).toEqual({
      data: {
        x: 1,
        y: ['array'],
        limit: '6',
      },
      headers: 'headers',
    });
    expect(apiSearch.calls.argsFor(3)[0]).toEqual({
      data: { filters: {}, types: [], limit: '12' },
      headers: 'headers',
    });

    assertItemLists(stateActions.itemLists);
  });

  describe('Datasets', () => {
    let markdownDatasetsResponse: {};

    beforeEach(() => {
      markdownDatasetsResponse = { request1: 'url1', request2: 'url2' };
      spyOn(markdownDatasets, 'fetch').and.callFake(async (content, requestParams, options) => {
        expect(content).toBe('originalContent');
        expect(requestParams).toEqual(request.onlyHeaders());
        return Promise.resolve({ ...markdownDatasetsResponse, ...options.additionalDatasets });
      });
    });

    it('should request each dataset inside the content', async () => {
      const stateActions = await getPageAssets(request);
      expect(stateActions.datasets).toEqual(markdownDatasetsResponse);
    });

    describe('Extended datasets and data', () => {
      it('should request additional dataset queries and passed data', async () => {
        const stateActions = await getPageAssets(
          request,
          { customDataset: { url: 'someurl' } },
          { customData: { localData: 'from store' } }
        );

        expect(stateActions.datasets).toEqual({
          request1: 'url1',
          request2: 'url2',
          customDataset: { url: 'someurl' },
          customData: { localData: 'from store' },
        });
      });
    });
  });

  describe('Dynamic content', () => {
    const localDatasets = {
      entityData: {
        sharedId: 'mtpkxxe1uom',
        title: 'My entity',
        metadata: {
          my_text_property: [
            {
              value: 'some text',
              displayValue: 'some text',
            },
          ],
          numericValue: [{ value: 1993, displayValue: 1993 }],
          a_date: [{ value: 747198000, displayValue: 'September 5, 1993' }],
          multiselect: [
            { value: '123fgfdcv', displayValue: 'Option 1' },
            { value: 'yjk56dfgd', displayValue: 'Option 2' },
          ],
          multiDateRange: [
            {
              value: {
                from: 1651968000,
                to: 1652486399,
              },
              displayValue: 'May 5, 2022 - May 5, 2023',
            },
            {
              value: {
                from: 1652572800,
                to: 1653091199,
              },
              displayValue: 'May 5, 2024 - May 5, 2025',
            },
          ],
        },
      },
      template: {
        name: 'Document',
        properties: [
          {
            _id: '6267e68226904c252518f914',
            label: 'Text',
            type: 'text',
            name: 'text',
            filter: true,
          },
        ],
      },
    };
    it('should parse the content and insert references to dataset', async () => {
      page.metadata.content =
        '<h1>${entity.metadata.my_text_property} from template ${template.name}</h1>';
      const assets = await getPageAssets(request, undefined, localDatasets);
      expect(assets.pageView.metadata.content).toBe('<h1>some text from template Document</h1>');
      expect(assets.errors).not.toBeDefined();
    });

    it.each`
      path                                              | result
      ${'entity.title'}                                 | ${'My entity'}
      ${'entity.sharedId'}                              | ${'mtpkxxe1uom'}
      ${'entity.metadata.a_date'}                       | ${747198000}
      ${'entity.metadata.a_date.value'}                 | ${747198000}
      ${'entity.metadata.a_date.displayValue'}          | ${'September 5, 1993'}
      ${'entity.metadata.multiselect'}                  | ${'123fgfdcv'}
      ${'entity.metadata.multiselect[0]'}               | ${'123fgfdcv'}
      ${'entity.metadata.multiselect[1]'}               | ${'yjk56dfgd'}
      ${'entity.metadata.multiselect.displayValue'}     | ${'Option 1'}
      ${'entity.metadata.multiselect[0].displayValue'}  | ${'Option 1'}
      ${'entity.metadata.multiselect[0].value'}         | ${'123fgfdcv'}
      ${'entity.metadata.multiselect[1].displayValue'}  | ${'Option 2'}
      ${'entity.metadata.multiDateRange'}               | ${{ from: 1651968000, to: 1652486399 }}
      ${'entity.metadata.multiDateRange[0].value.from'} | ${1651968000}
    `('should work for entity path $path', async ({ path, result }) => {
      page.metadata.content = `<p>My dynamic path results is: \${${path}}</p>`;
      const assets = await getPageAssets(request, undefined, localDatasets);
      expect(assets.pageView.metadata.content).toBe(`<p>My dynamic path results is: ${result}</p>`);
    });

    it('should ignore references if they are not part of a dataset', async () => {
      page.metadata.content = '<h1>${entity.sharedId} from template ${template.metadata}</h1>';
      const assets = await getPageAssets(request, undefined, localDatasets);
      expect(assets.pageView.metadata.content).toBe(
        '<h1>mtpkxxe1uom from template ${template.metadata}</h1>'
      );
      expect(assets.errors).toEqual(
        'The following expressions are not valid properties:\n ${template.metadata}'
      );
    });
  });
});
