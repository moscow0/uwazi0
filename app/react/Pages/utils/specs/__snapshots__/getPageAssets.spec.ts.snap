// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPageAssets should request page for view 1`] = `
Object {
  "datasets": Object {},
  "itemLists": Array [
    Object {
      "items": Array [
        "resultsFor:0",
      ],
      "options": Object {},
      "params": "?q=(a:1,b:2)",
    },
    Object {
      "items": Array [
        "resultsFor:1",
      ],
      "options": Object {
        "limit": 9,
      },
      "params": "",
    },
    Object {
      "items": Array [
        "resultsFor:2",
      ],
      "options": Object {
        "limit": 0,
      },
      "params": "?q=(x:1,y:!(%27array%27),limit:24)",
    },
    Object {
      "items": Array [
        "resultsFor:3",
      ],
      "options": Object {
        "limit": 12,
      },
      "params": "?order=metadata.form&treatAs=number",
    },
  ],
  "pageView": Object {
    "_id": "abc2",
    "metadata": Object {
      "content": "originalContent",
    },
    "scriptRendered": false,
    "title": "Page 1",
  },
}
`;
